src/compiler/transformers/esnext.ts

```diff
@@ -16,9 +19,67 @@ namespace ts {
                 return node;
             }
             switch (node.kind) {
+                case SyntaxKind.BinaryExpression:
+                    const binaryExpression = <BinaryExpression>node;
+                    if (isLogicalOrCoalescingAssignmentOperator(binaryExpression.operatorToken.kind)) {
+                        return transformLogicalAssignmentOperators(binaryExpression);
+                    }
+                    // falls through
                 default:
                     return visitEachChild(node, visitor, context);
             }
         }
+
+        function transformLogicalAssignmentOperators(binaryExpression: BinaryExpression): VisitResult<Node> {
+            const operator = binaryExpression.operatorToken;
+            if (isCompoundAssignment(operator.kind) && isLogicalOrCoalescingAssignmentOperator(operator.kind)) {

```

Comment:
Do you even need these conditions here since `isLogicalOrCoalescingAssignmentOperator` is checked before the function is called? If anything I would just move this to a `Debug.assert` at the top of the function. Alternatively, you could change the type of `binaryExpression` to `AssignmentExpression<SyntaxKind.AmpersandAmpersandEqualsToken | SyntaxKind.BarBarEqualsToken | SyntaxKind.QuestionQuestionEqualsToken>` to enforce the condition at compile time instead of at runtime.